GitUI/CommandsDialogs/BrowseDialog/FormCommitCount.cs

```diff
@@ -1,96 +1,141 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows.Forms;
 using GitCommands;
 using GitCommands.Statistics;
+using Microsoft.VisualStudio.Threading;
 
 namespace GitUI.CommandsDialogs.BrowseDialog
 {
     public partial class FormCommitCount : GitModuleForm
     {
+        private CancellationTokenSource _backgroundLoaderTokenSource = new CancellationTokenSource();
+
         public FormCommitCount(GitUICommands commands)
             : base(commands)
         {
             InitializeComponent();
             Loading.Image = Properties.Resources.loadingpanel;
             Translate();
-        }
 
-        private void FormCommitCountLoad(object sender, EventArgs e)
-        {
-            FetchData();
+            Load += delegate { FetchData(); };
+            cbIncludeSubmodules.CheckedChanged += delegate { FetchData(); };
         }
 
-        private void cbIncludeSubmodules_CheckedChanged(object sender, EventArgs e)
+        private void FetchData()
         {
-            FetchData();
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            _backgroundLoaderTokenSource.Cancel();
+            _backgroundLoaderTokenSource.Dispose();
+            _backgroundLoaderTokenSource = new CancellationTokenSource();
+            var token = _backgroundLoaderTokenSource.Token;
+
+            ThreadHelper.JoinableTaskFactory.RunAsync(
+                async () =>
+                {
+                    Loading.Visible = true;
+
+                    CommitCount.Text = "";
+
+                    var includeSubmodules = cbIncludeSubmodules.Checked;
+
+                    await TaskScheduler.Default;
+
+                    var text = GenerateText(Module, includeSubmodules, token);
```

Comment:
Passing `Module` like this is not safe, if `Module` is reassigned in mean time the operation may execute against an incorrect git repo.
It should be done something like this:
https://github.com/gitextensions/gitextensions/blob/d22a21ac1ed21a5f784cd518eb003152e21f6372/GitUI/CommandsDialogs/FormBrowse.cs#L2556-L2563

This is one of the reasons why I am insistent on splitting state and behaviours mixed together in `GitModule`.

Reply:
💭 I'm not sure this is correct in the context of the current form. The module for this form is initialized in the constructor and never changed.

Reply:
Yes, you're right.
This is one of those inconsistencies we need to address across the codebase